From 1becb923e5c199fff626a4a4c01f908c97e989af Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?M=C3=A5rten=20Kongstad?= <marten.kongstad@sony.com>
Date: Thu, 1 Jun 2017 09:37:01 +0200
Subject: [PATCH 12/15] OMS: add support for user-installed overlays

Teach the package manager to look for overlay packages everywhere, not
just in /vendor/overlay. Implement the necessary callbacks in the
overlay manager (onOverlayPackage.*).

Also, update the overlay manager unit tests to install its own overlays
as part of the setup rather than relying on the user to have done so
before.

Note: this commit allows any overlay package to be used which is a
security risk. A future commit will harden security by restricting which
overlay packages are allowed to take effect.

Test: bit FrameworksServicesTests:com.android.server.om.{OverlayManagerSettingsTests,OverlayManagerTests,RuntimeResourceOverlayTests}
Change-Id: Ib4556b0f65cc27772bc2efb6063ff1963811de3d
---
 core/java/android/content/om/OverlayInfo.java      |  22 +++
 core/java/android/content/pm/PackageParser.java    |   4 -
 .../server/om/OverlayManagerServiceImpl.java       | 162 +++++++++++++++------
 .../android/server/pm/PackageManagerService.java   |  35 ++---
 .../tests/servicestests/prepare-overlay-tests.sh   |  10 --
 .../com/android/server/om/OverlayManagerTests.java |  85 ++++++++++-
 .../src/com/android/server/om/OverlayUtils.java    |  37 ++++-
 .../src/com/android/server/om/PackageUtils.java    |   6 +-
 .../server/om/RuntimeResourceOverlayTests.java     |  69 +++++----
 .../com/android/server/pm/PackageParserTest.java   |   2 -
 .../test-apks/app_overlay_1/AndroidManifest.xml    |   4 +-
 .../test-apks/app_overlay_1_v2/AndroidManifest.xml |   6 +
 .../servicestests/test-apks/app_overlay_1_v2/build |  24 +++
 .../app_overlay_1_v2/res/values/values.xml         |   4 +
 14 files changed, 353 insertions(+), 117 deletions(-)
 delete mode 100644 services/tests/servicestests/prepare-overlay-tests.sh
 create mode 100644 services/tests/servicestests/test-apks/app_overlay_1_v2/AndroidManifest.xml
 create mode 100644 services/tests/servicestests/test-apks/app_overlay_1_v2/build
 create mode 100644 services/tests/servicestests/test-apks/app_overlay_1_v2/res/values/values.xml

diff --git a/core/java/android/content/om/OverlayInfo.java b/core/java/android/content/om/OverlayInfo.java
index 91bed74185b..47e12b232db 100644
--- a/core/java/android/content/om/OverlayInfo.java
+++ b/core/java/android/content/om/OverlayInfo.java
@@ -35,6 +35,18 @@ public final class OverlayInfo implements Parcelable {
      */
     public static final int STATE_UNKNOWN = -1;
 
+    /**
+     * The target package is currently being upgraded; the state will change
+     * once the package installation has finished.
+     */
+    public static final int STATE_TARGET_UPGRADING = 4;
+
+    /**
+     * The overlay package is currently being upgraded; the state will change
+     * once the package installation has finished.
+     */
+    public static final int STATE_OVERLAY_UPGRADING = 5;
+
     /**
      * The target package of the overlay is not installed. The overlay cannot be enabled.
      */
@@ -86,6 +98,8 @@ public final class OverlayInfo implements Parcelable {
     /**
      * The state of this OverlayInfo as defined by the STATE_* constants in this class.
      *
+     * @see #STATE_TARGET_UPGRADING
+     * @see #STATE_OVERLAY_UPGRADING
      * @see #STATE_MISSING_TARGET
      * @see #STATE_NO_IDMAP
      * @see #STATE_DISABLED
@@ -142,6 +156,8 @@ public final class OverlayInfo implements Parcelable {
         }
         switch (state) {
             case STATE_UNKNOWN:
+            case STATE_TARGET_UPGRADING:
+            case STATE_OVERLAY_UPGRADING:
             case STATE_MISSING_TARGET:
             case STATE_NO_IDMAP:
             case STATE_DISABLED:
@@ -200,6 +216,8 @@ public final class OverlayInfo implements Parcelable {
      * Translate a state to a human readable string. Only intended for
      * debugging purposes.
      *
+     * @see #STATE_TARGET_UPGRADING
+     * @see #STATE_OVERLAY_UPGRADING
      * @see #STATE_MISSING_TARGET
      * @see #STATE_NO_IDMAP
      * @see #STATE_DISABLED
@@ -211,6 +229,10 @@ public final class OverlayInfo implements Parcelable {
         switch (state) {
             case STATE_UNKNOWN:
                 return "STATE_UNKNOWN";
+            case STATE_TARGET_UPGRADING:
+                return "STATE_TARGET_UPGRADING";
+            case STATE_OVERLAY_UPGRADING:
+                return "STATE_OVERLAY_UPGRADING";
             case STATE_MISSING_TARGET:
                 return "STATE_MISSING_TARGET";
             case STATE_NO_IDMAP:
diff --git a/core/java/android/content/pm/PackageParser.java b/core/java/android/content/pm/PackageParser.java
index 7b4e69059e9..298ac9c6fad 100644
--- a/core/java/android/content/pm/PackageParser.java
+++ b/core/java/android/content/pm/PackageParser.java
@@ -826,7 +826,6 @@ public class PackageParser {
     public final static int PARSE_IS_SYSTEM_DIR = 1<<6;
     public final static int PARSE_IS_PRIVILEGED = 1<<7;
     public final static int PARSE_COLLECT_CERTIFICATES = 1<<8;
-    public final static int PARSE_TRUSTED_OVERLAY = 1<<9;
     public final static int PARSE_ENFORCE_CODE = 1<<10;
     /** @deprecated remove when fixing b/34761192 */
     @Deprecated
@@ -5950,7 +5949,6 @@ public class PackageParser {
         public String mOverlayTarget;
         public int mOverlayPriority;
         public boolean mIsStaticOverlay;
-        public boolean mTrustedOverlay;
 
         /**
          * Data used to feed the KeySetManagerService
@@ -6442,7 +6440,6 @@ public class PackageParser {
             mOverlayTarget = dest.readString();
             mOverlayPriority = dest.readInt();
             mIsStaticOverlay = (dest.readInt() == 1);
-            mTrustedOverlay = (dest.readInt() == 1);
             mSigningKeys = (ArraySet<PublicKey>) dest.readArraySet(boot);
             mUpgradeKeySets = (ArraySet<String>) dest.readArraySet(boot);
 
@@ -6565,7 +6562,6 @@ public class PackageParser {
             dest.writeString(mOverlayTarget);
             dest.writeInt(mOverlayPriority);
             dest.writeInt(mIsStaticOverlay ? 1 : 0);
-            dest.writeInt(mTrustedOverlay ? 1 : 0);
             dest.writeArraySet(mSigningKeys);
             dest.writeArraySet(mUpgradeKeySets);
             writeKeySetMapping(dest, mKeySetMapping);
diff --git a/services/core/java/com/android/server/om/OverlayManagerServiceImpl.java b/services/core/java/com/android/server/om/OverlayManagerServiceImpl.java
index 5e9dd6a3c03..7146c8cf8f3 100644
--- a/services/core/java/com/android/server/om/OverlayManagerServiceImpl.java
+++ b/services/core/java/com/android/server/om/OverlayManagerServiceImpl.java
@@ -20,6 +20,8 @@ import static android.content.om.OverlayInfo.STATE_DISABLED;
 import static android.content.om.OverlayInfo.STATE_ENABLED;
 import static android.content.om.OverlayInfo.STATE_MISSING_TARGET;
 import static android.content.om.OverlayInfo.STATE_NO_IDMAP;
+import static android.content.om.OverlayInfo.STATE_OVERLAY_UPGRADING;
+import static android.content.om.OverlayInfo.STATE_TARGET_UPGRADING;
 
 import static com.android.server.om.OverlayManagerService.DEBUG;
 import static com.android.server.om.OverlayManagerService.TAG;
@@ -50,6 +52,10 @@ import java.util.Set;
  * @see OverlayManagerService
  */
 final class OverlayManagerServiceImpl {
+    // Flags to use in conjunction with updateState.
+    private static final int FLAG_TARGET_IS_UPGRADING = 1<<0;
+    private static final int FLAG_OVERLAY_IS_UPGRADING = 1<<1;
+
     private final PackageManagerHelper mPackageManager;
     private final IdmapManager mIdmapManager;
     private final OverlayManagerSettings mSettings;
@@ -123,9 +129,7 @@ final class OverlayManagerServiceImpl {
             }
 
             try {
-                final PackageInfo targetPackage =
-                        mPackageManager.getPackageInfo(overlayPackage.overlayTarget, newUserId);
-                updateState(targetPackage, overlayPackage, newUserId);
+                updateState(overlayPackage.overlayTarget, overlayPackage.packageName, newUserId, 0);
             } catch (OverlayManagerSettings.BadKeyException e) {
                 Slog.e(TAG, "failed to update settings", e);
                 mSettings.remove(overlayPackage.packageName, newUserId);
@@ -168,8 +172,7 @@ final class OverlayManagerServiceImpl {
             Slog.d(TAG, "onTargetPackageAdded packageName=" + packageName + " userId=" + userId);
         }
 
-        final PackageInfo targetPackage = mPackageManager.getPackageInfo(packageName, userId);
-        if (updateAllOverlaysForTarget(packageName, userId, targetPackage)) {
+        if (updateAllOverlaysForTarget(packageName, userId, 0)) {
             mListener.onOverlaysChanged(packageName, userId);
         }
     }
@@ -179,8 +182,7 @@ final class OverlayManagerServiceImpl {
             Slog.d(TAG, "onTargetPackageChanged packageName=" + packageName + " userId=" + userId);
         }
 
-        final PackageInfo targetPackage = mPackageManager.getPackageInfo(packageName, userId);
-        if (updateAllOverlaysForTarget(packageName, userId, targetPackage)) {
+        if (updateAllOverlaysForTarget(packageName, userId, 0)) {
             mListener.onOverlaysChanged(packageName, userId);
         }
     }
@@ -190,7 +192,7 @@ final class OverlayManagerServiceImpl {
             Slog.d(TAG, "onTargetPackageUpgrading packageName=" + packageName + " userId=" + userId);
         }
 
-        if (updateAllOverlaysForTarget(packageName, userId, null)) {
+        if (updateAllOverlaysForTarget(packageName, userId, FLAG_TARGET_IS_UPGRADING)) {
             mListener.onOverlaysChanged(packageName, userId);
         }
     }
@@ -200,8 +202,7 @@ final class OverlayManagerServiceImpl {
             Slog.d(TAG, "onTargetPackageUpgraded packageName=" + packageName + " userId=" + userId);
         }
 
-        final PackageInfo targetPackage = mPackageManager.getPackageInfo(packageName, userId);
-        if (updateAllOverlaysForTarget(packageName, userId, targetPackage)) {
+        if (updateAllOverlaysForTarget(packageName, userId, 0)) {
             mListener.onOverlaysChanged(packageName, userId);
         }
     }
@@ -211,7 +212,7 @@ final class OverlayManagerServiceImpl {
             Slog.d(TAG, "onTargetPackageRemoved packageName=" + packageName + " userId=" + userId);
         }
 
-        if (updateAllOverlaysForTarget(packageName, userId, null)) {
+        if (updateAllOverlaysForTarget(packageName, userId, 0)) {
             mListener.onOverlaysChanged(packageName, userId);
         }
     }
@@ -219,10 +220,10 @@ final class OverlayManagerServiceImpl {
     /**
      * Returns true if the settings were modified for this target.
      */
-    private boolean updateAllOverlaysForTarget(@NonNull final String packageName, final int userId,
-            @Nullable final PackageInfo targetPackage) {
+    private boolean updateAllOverlaysForTarget(@NonNull final String targetPackageName,
+            final int userId, final int flags) {
         boolean modified = false;
-        final List<OverlayInfo> ois = mSettings.getOverlaysForTarget(packageName, userId);
+        final List<OverlayInfo> ois = mSettings.getOverlaysForTarget(targetPackageName, userId);
         final int N = ois.size();
         for (int i = 0; i < N; i++) {
             final OverlayInfo oi = ois.get(i);
@@ -232,7 +233,7 @@ final class OverlayManagerServiceImpl {
                 removeIdmapIfPossible(oi);
             } else {
                 try {
-                    modified |= updateState(targetPackage, overlayPackage, userId);
+                    modified |= updateState(targetPackageName, oi.packageName, userId, flags);
                 } catch (OverlayManagerSettings.BadKeyException e) {
                     Slog.e(TAG, "failed to update settings", e);
                     modified |= mSettings.remove(oi.packageName, userId);
@@ -254,14 +255,11 @@ final class OverlayManagerServiceImpl {
             return;
         }
 
-        final PackageInfo targetPackage =
-                mPackageManager.getPackageInfo(overlayPackage.overlayTarget, userId);
-
         mSettings.init(packageName, userId, overlayPackage.overlayTarget,
                 overlayPackage.applicationInfo.getBaseCodePath(), overlayPackage.isStaticOverlay,
                 overlayPackage.overlayPriority);
         try {
-            if (updateState(targetPackage, overlayPackage, userId)) {
+            if (updateState(overlayPackage.overlayTarget, packageName, userId, 0)) {
                 mListener.onOverlaysChanged(overlayPackage.overlayTarget, userId);
             }
         } catch (OverlayManagerSettings.BadKeyException e) {
@@ -271,19 +269,80 @@ final class OverlayManagerServiceImpl {
     }
 
     void onOverlayPackageChanged(@NonNull final String packageName, final int userId) {
-        Slog.wtf(TAG, "onOverlayPackageChanged called, but only pre-installed overlays supported");
+        if (DEBUG) {
+            Slog.d(TAG, "onOverlayPackageChanged packageName=" + packageName + " userId=" + userId);
+        }
+
+        try {
+            final OverlayInfo oi = mSettings.getOverlayInfo(packageName, userId);
+            if (updateState(oi.targetPackageName, packageName, userId, 0)) {
+                mListener.onOverlaysChanged(oi.targetPackageName, userId);
+            }
+        } catch (OverlayManagerSettings.BadKeyException e) {
+            Slog.e(TAG, "failed to update settings", e);
+        }
     }
 
     void onOverlayPackageUpgrading(@NonNull final String packageName, final int userId) {
-        Slog.wtf(TAG, "onOverlayPackageUpgrading called, but only pre-installed overlays supported");
+        if (DEBUG) {
+            Slog.d(TAG, "onOverlayPackageUpgrading packageName=" + packageName + " userId=" + userId);
+        }
+
+        try {
+            final OverlayInfo oi = mSettings.getOverlayInfo(packageName, userId);
+            if (updateState(oi.targetPackageName, packageName, userId, FLAG_OVERLAY_IS_UPGRADING)) {
+                removeIdmapIfPossible(oi);
+                mListener.onOverlaysChanged(oi.targetPackageName, userId);
+            }
+        } catch (OverlayManagerSettings.BadKeyException e) {
+            Slog.e(TAG, "failed to update settings", e);
+        }
     }
 
     void onOverlayPackageUpgraded(@NonNull final String packageName, final int userId) {
-        Slog.wtf(TAG, "onOverlayPackageUpgraded called, but only pre-installed overlays supported");
+        if (DEBUG) {
+            Slog.d(TAG, "onOverlayPackageUpgraded packageName=" + packageName + " userId=" + userId);
+        }
+
+        final PackageInfo pkg = mPackageManager.getPackageInfo(packageName, userId);
+        if (pkg == null) {
+            Slog.w(TAG, "overlay package " + packageName + " was upgraded, but couldn't be found");
+            onOverlayPackageRemoved(packageName, userId);
+            return;
+        }
+
+        try {
+            final OverlayInfo oldOi = mSettings.getOverlayInfo(packageName, userId);
+            if (oldOi.targetPackageName.equals(pkg.overlayTarget)) {
+                updateState(oldOi.targetPackageName, packageName, userId, 0);
+                mListener.onOverlaysChanged(pkg.overlayTarget, userId);
+            } else {
+                mSettings.remove(packageName, userId);
+                mSettings.init(packageName, userId, pkg.overlayTarget,
+                        pkg.applicationInfo.getBaseCodePath(), pkg.isStaticOverlay,
+                        pkg.overlayPriority);
+                updateState(pkg.overlayTarget, packageName, userId, 0);
+                mListener.onOverlaysChanged(pkg.overlayTarget, userId);
+            }
+        } catch (OverlayManagerSettings.BadKeyException e) {
+            Slog.e(TAG, "failed to update settings", e);
+        }
     }
 
     void onOverlayPackageRemoved(@NonNull final String packageName, final int userId) {
-        Slog.wtf(TAG, "onOverlayPackageRemoved called, but only pre-installed overlays supported");
+        if (DEBUG) {
+            Slog.d(TAG, "onOverlayPackageRemoved packageName=" + packageName + " userId=" + userId);
+        }
+
+        try {
+            final OverlayInfo oi = mSettings.getOverlayInfo(packageName, userId);
+            if (mSettings.remove(packageName, userId)) {
+                removeIdmapIfPossible(oi);
+                mListener.onOverlaysChanged(oi.targetPackageName, userId);
+            }
+        } catch (OverlayManagerSettings.BadKeyException e) {
+            Slog.e(TAG, "failed to update settings", e);
+        }
     }
 
     OverlayInfo getOverlayInfo(@NonNull final String packageName, final int userId) {
@@ -322,10 +381,8 @@ final class OverlayManagerServiceImpl {
 
         try {
             final OverlayInfo oi = mSettings.getOverlayInfo(packageName, userId);
-            final PackageInfo targetPackage =
-                    mPackageManager.getPackageInfo(oi.targetPackageName, userId);
             boolean modified = mSettings.setEnabled(packageName, userId, enable);
-            modified |= updateState(targetPackage, overlayPackage, userId);
+            modified |= updateState(oi.targetPackageName, oi.packageName, userId, 0);
 
             if (modified) {
                 mListener.onOverlaysChanged(oi.targetPackageName, userId);
@@ -348,10 +405,9 @@ final class OverlayManagerServiceImpl {
 
         try {
             final OverlayInfo oi = mSettings.getOverlayInfo(packageName, userId);
-            final PackageInfo targetPackage =
-                    mPackageManager.getPackageInfo(oi.targetPackageName, userId);
+            final String targetPackageName = oi.targetPackageName;
 
-            List<OverlayInfo> allOverlays = getOverlayInfosForTarget(oi.targetPackageName, userId);
+            List<OverlayInfo> allOverlays = getOverlayInfosForTarget(targetPackageName, userId);
 
             boolean modified = false;
 
@@ -373,15 +429,15 @@ final class OverlayManagerServiceImpl {
 
                 // Disable the overlay.
                 modified |= mSettings.setEnabled(disabledOverlayPackageName, userId, false);
-                modified |= updateState(targetPackage, disabledOverlayPackageInfo, userId);
+                modified |= updateState(targetPackageName, disabledOverlayPackageName, userId, 0);
             }
 
             // Enable the selected overlay.
             modified |= mSettings.setEnabled(packageName, userId, true);
-            modified |= updateState(targetPackage, overlayPackage, userId);
+            modified |= updateState(targetPackageName, packageName, userId, 0);
 
             if (modified) {
-                mListener.onOverlaysChanged(oi.targetPackageName, userId);
+                mListener.onOverlaysChanged(targetPackageName, userId);
             }
             return true;
         } catch (OverlayManagerSettings.BadKeyException e) {
@@ -481,35 +537,55 @@ final class OverlayManagerServiceImpl {
     /**
      * Returns true if the settings/state was modified, false otherwise.
      */
-    private boolean updateState(@Nullable final PackageInfo targetPackage,
-            @NonNull final PackageInfo overlayPackage, final int userId)
+    private boolean updateState(@NonNull final String targetPackageName,
+            @NonNull final String overlayPackageName, final int userId, final int flags)
             throws OverlayManagerSettings.BadKeyException {
+
+        final PackageInfo targetPackage = mPackageManager.getPackageInfo(targetPackageName, userId);
+        final PackageInfo overlayPackage = mPackageManager.getPackageInfo(overlayPackageName, userId);
+
         // Static RROs targeting to "android", ie framework-res.apk, are handled by native layers.
-        if (targetPackage != null &&
-                !("android".equals(targetPackage.packageName) && overlayPackage.isStaticOverlay)) {
+        if (targetPackage != null && overlayPackage != null &&
+                !("android".equals(targetPackageName) && overlayPackage.isStaticOverlay)) {
             mIdmapManager.createIdmap(targetPackage, overlayPackage, userId);
         }
 
-        boolean modified = mSettings.setBaseCodePath(overlayPackage.packageName, userId,
-                overlayPackage.applicationInfo.getBaseCodePath());
+        boolean modified = false;
+        if (overlayPackage != null) {
+            modified |= mSettings.setBaseCodePath(overlayPackageName, userId,
+                    overlayPackage.applicationInfo.getBaseCodePath());
+        }
 
-        final int currentState = mSettings.getState(overlayPackage.packageName, userId);
-        final int newState = calculateNewState(targetPackage, overlayPackage, userId);
+        final int currentState = mSettings.getState(overlayPackageName, userId);
+        final int newState = calculateNewState(targetPackage, overlayPackage, userId, flags);
         if (currentState != newState) {
             if (DEBUG) {
                 Slog.d(TAG, String.format("%s:%d: %s -> %s",
-                            overlayPackage.packageName, userId,
+                            overlayPackageName, userId,
                             OverlayInfo.stateToString(currentState),
                             OverlayInfo.stateToString(newState)));
             }
-            modified |= mSettings.setState(overlayPackage.packageName, userId, newState);
+            modified |= mSettings.setState(overlayPackageName, userId, newState);
         }
         return modified;
     }
 
     private int calculateNewState(@Nullable final PackageInfo targetPackage,
-            @NonNull final PackageInfo overlayPackage, final int userId)
+            @Nullable final PackageInfo overlayPackage, final int userId, final int flags)
         throws OverlayManagerSettings.BadKeyException {
+        if ((flags & FLAG_TARGET_IS_UPGRADING) != 0) {
+            return STATE_TARGET_UPGRADING;
+        }
+
+        if ((flags & FLAG_OVERLAY_IS_UPGRADING) != 0) {
+            return STATE_OVERLAY_UPGRADING;
+        }
+
+        // assert expectation on overlay package: can only be null if the flags are used
+        if (DEBUG && overlayPackage == null) {
+            throw new IllegalArgumentException("null overlay package not compatible with no flags");
+        }
+
         if (targetPackage == null) {
             return STATE_MISSING_TARGET;
         }
diff --git a/services/core/java/com/android/server/pm/PackageManagerService.java b/services/core/java/com/android/server/pm/PackageManagerService.java
index c98243ee13f..047af3a632e 100644
--- a/services/core/java/com/android/server/pm/PackageManagerService.java
+++ b/services/core/java/com/android/server/pm/PackageManagerService.java
@@ -440,19 +440,18 @@ public class PackageManagerService extends IPackageManager.Stub
     static final int SCAN_NEW_INSTALL = 1<<4;
     static final int SCAN_UPDATE_TIME = 1<<5;
     static final int SCAN_BOOTING = 1<<6;
-    static final int SCAN_TRUSTED_OVERLAY = 1<<7;
-    static final int SCAN_DELETE_DATA_ON_FAILURES = 1<<8;
-    static final int SCAN_REPLACING = 1<<9;
-    static final int SCAN_REQUIRE_KNOWN = 1<<10;
-    static final int SCAN_MOVE = 1<<11;
-    static final int SCAN_INITIAL = 1<<12;
-    static final int SCAN_CHECK_ONLY = 1<<13;
-    static final int SCAN_DONT_KILL_APP = 1<<14;
-    static final int SCAN_IGNORE_FROZEN = 1<<15;
-    static final int SCAN_FIRST_BOOT_OR_UPGRADE = 1<<16;
-    static final int SCAN_AS_INSTANT_APP = 1<<17;
-    static final int SCAN_AS_FULL_APP = 1<<18;
-    static final int SCAN_AS_VIRTUAL_PRELOAD = 1<<19;
+    static final int SCAN_DELETE_DATA_ON_FAILURES = 1<<7;
+    static final int SCAN_REPLACING = 1<<8;
+    static final int SCAN_REQUIRE_KNOWN = 1<<9;
+    static final int SCAN_MOVE = 1<<10;
+    static final int SCAN_INITIAL = 1<<11;
+    static final int SCAN_CHECK_ONLY = 1<<12;
+    static final int SCAN_DONT_KILL_APP = 1<<13;
+    static final int SCAN_IGNORE_FROZEN = 1<<14;
+    static final int SCAN_FIRST_BOOT_OR_UPGRADE = 1<<15;
+    static final int SCAN_AS_INSTANT_APP = 1<<16;
+    static final int SCAN_AS_FULL_APP = 1<<17;
+    static final int SCAN_AS_VIRTUAL_PRELOAD = 1<<18;
     /** Should not be with the scan flags */
     static final int FLAGS_REMOVE_CHATTY = 1<<31;
 
@@ -2609,12 +2608,9 @@ public class PackageManagerService extends IPackageManager.Stub
             }
 
             // Collect vendor overlay packages. (Do this before scanning any apps.)
-            // For security and version matching reason, only consider
-            // overlay packages if they reside in the right directory.
             scanDirTracedLI(new File(VENDOR_OVERLAY_DIR), mDefParseFlags
                     | PackageParser.PARSE_IS_SYSTEM
-                    | PackageParser.PARSE_IS_SYSTEM_DIR
-                    | PackageParser.PARSE_TRUSTED_OVERLAY, scanFlags | SCAN_TRUSTED_OVERLAY, 0);
+                    | PackageParser.PARSE_IS_SYSTEM_DIR, scanFlags, 0);
 
             mParallelPackageParserCallback.findStaticOverlayPackages();
 
@@ -9226,10 +9222,6 @@ public class PackageManagerService extends IPackageManager.Stub
         pp.setDisplayMetrics(mMetrics);
         pp.setCallback(mPackageParserCallback);
 
-        if ((scanFlags & SCAN_TRUSTED_OVERLAY) != 0) {
-            parseFlags |= PackageParser.PARSE_TRUSTED_OVERLAY;
-        }
-
         Trace.traceBegin(TRACE_TAG_PACKAGE_MANAGER, "parsePackage");
         final PackageParser.Package pkg;
         try {
@@ -11217,7 +11209,6 @@ public class PackageManagerService extends IPackageManager.Stub
             pkg.applicationInfo.privateFlags &=
                     ~ApplicationInfo.PRIVATE_FLAG_DIRECT_BOOT_AWARE;
         }
-        pkg.mTrustedOverlay = (policyFlags&PackageParser.PARSE_TRUSTED_OVERLAY) != 0;
 
         if ((policyFlags&PackageParser.PARSE_IS_PRIVILEGED) != 0) {
             pkg.applicationInfo.privateFlags |= ApplicationInfo.PRIVATE_FLAG_PRIVILEGED;
diff --git a/services/tests/servicestests/prepare-overlay-tests.sh b/services/tests/servicestests/prepare-overlay-tests.sh
deleted file mode 100644
index eff8d733aac..00000000000
--- a/services/tests/servicestests/prepare-overlay-tests.sh
+++ /dev/null
@@ -1,10 +0,0 @@
-pkgs=''
-pkgs+='res/raw/app_overlay_1 '
-pkgs+='res/raw/app_overlay_2 '
-pkgs+='res/raw/some_other_app_overlay '
-pkgs+='res/raw/system_overlay_1 '
-pkgs+='res/raw/system_overlay_2 '
-
-for pkg in $pkgs; do
-    adb push $pkg /vendor/overlay/$(basename $pkg).apk
-done
diff --git a/services/tests/servicestests/src/com/android/server/om/OverlayManagerTests.java b/services/tests/servicestests/src/com/android/server/om/OverlayManagerTests.java
index 6f8c3cb9676..954cde8bee6 100644
--- a/services/tests/servicestests/src/com/android/server/om/OverlayManagerTests.java
+++ b/services/tests/servicestests/src/com/android/server/om/OverlayManagerTests.java
@@ -3,6 +3,9 @@ package com.android.server.om;
 import static android.content.om.OverlayInfo.STATE_DISABLED;
 import static android.content.om.OverlayInfo.STATE_ENABLED;
 import static android.content.om.OverlayInfo.STATE_MISSING_TARGET;
+import static android.content.om.OverlayInfo.STATE_NO_IDMAP;
+import static android.content.om.OverlayInfo.STATE_OVERLAY_UPGRADING;
+import static android.content.om.OverlayInfo.STATE_TARGET_UPGRADING;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
@@ -159,7 +162,7 @@ public class OverlayManagerTests {
         assertState(STATE_ENABLED, OVERLAY, USER);
 
         beginUpgradeTargetPackage(TARGET, USER);
-        assertState(STATE_MISSING_TARGET, OVERLAY, USER);
+        assertState(STATE_TARGET_UPGRADING, OVERLAY, USER);
 
         endUpgradeTargetPackage(TARGET, USER);
         assertState(STATE_ENABLED, OVERLAY, USER);
@@ -191,6 +194,62 @@ public class OverlayManagerTests {
         assertEquals(0, mListener.count);
     }
 
+    // tests: overlay installation and removal
+
+    @Test
+    public void testUninstallOverlay() throws Exception {
+        assertNull(mImpl.getOverlayInfo(OVERLAY, USER));
+
+        installOverlayPackage(OVERLAY, TARGET, USER, false);
+        assertNotNull(mImpl.getOverlayInfo(OVERLAY, USER));
+
+        uninstallOverlayPackage(OVERLAY, USER);
+        assertNull(mImpl.getOverlayInfo(OVERLAY, USER));
+    }
+
+    @Test
+    public void testUpgradeOverlay() throws Exception {
+        installOverlayPackage(OVERLAY, TARGET, USER, true);
+        installTargetPackage(TARGET, USER);
+        mImpl.setEnabled(OVERLAY, true, USER);
+        assertState(STATE_ENABLED, OVERLAY, USER);
+
+        beginUpgradeOverlayPackage(OVERLAY, USER);
+        assertState(STATE_OVERLAY_UPGRADING, OVERLAY, USER);
+
+        endUpgradeOverlayPackage(OVERLAY, TARGET, USER, true);
+        assertState(STATE_ENABLED, OVERLAY, USER);
+
+        beginUpgradeOverlayPackage(OVERLAY, USER);
+        assertState(STATE_OVERLAY_UPGRADING, OVERLAY, USER);
+
+        endUpgradeOverlayPackage(OVERLAY, TARGET, USER, false);
+        assertState(STATE_NO_IDMAP, OVERLAY, USER);
+    }
+
+    @Test
+    public void testUpgradeSneakyOverlay() throws Exception {
+        final String otherTarget = "some.other.target";
+        installTargetPackage(otherTarget, USER);
+
+        installOverlayPackage(OVERLAY, TARGET, USER, true);
+        installTargetPackage(TARGET, USER);
+        mImpl.setEnabled(OVERLAY, true, USER);
+        assertState(STATE_ENABLED, OVERLAY, USER);
+
+        beginUpgradeOverlayPackage(OVERLAY, USER);
+        assertState(STATE_OVERLAY_UPGRADING, OVERLAY, USER);
+
+        // changing the overlay's target as part of an upgrade should be the
+        // same as uninstalling the overlay and installing the new version;
+        // especially the OverlayInfo's target should be updated and it should
+        // not be enabled
+        endUpgradeOverlayPackage(OVERLAY, otherTarget, USER, true);
+        assertState(STATE_DISABLED, OVERLAY, USER);
+        final OverlayInfo oi = mImpl.getOverlayInfo(OVERLAY, USER);
+        assertEquals(otherTarget, oi.targetPackageName);
+    }
+
     // helper methods
 
     private void assertState(int expected, final String overlayPackageName, int userId) {
@@ -249,15 +308,29 @@ public class OverlayManagerTests {
         mImpl.onOverlayPackageAdded(packageName, userId);
     }
 
-    private void upgradeOverlayPackage(String packageName, String targetPackageName, int userId,
+    private void beginUpgradeOverlayPackage(String packageName, int userId) {
+        if (mState.select(packageName, userId) == null) {
+            throw new IllegalStateException("package not installed");
+        }
+        mState.add(packageName, null, userId, false);
+        mImpl.onOverlayPackageUpgrading(packageName, userId);
+    }
+
+    private void endUpgradeOverlayPackage(String packageName, String targetPackageName, int userId,
             boolean canCreateIdmap) {
-        // implement this when adding support for downloadable overlays
-        throw new IllegalArgumentException("not implemented");
+        if (mState.select(packageName, userId) == null) {
+            throw new IllegalStateException("package not installed");
+        }
+        mState.add(packageName, targetPackageName, userId, canCreateIdmap);
+        mImpl.onOverlayPackageUpgraded(packageName, userId);
     }
 
     private void uninstallOverlayPackage(String packageName, int userId) {
-        // implement this when adding support for downloadable overlays
-        throw new IllegalArgumentException("not implemented");
+        if (mState.select(packageName, userId) == null) {
+            throw new IllegalStateException("package not installed");
+        }
+        mState.remove(packageName, userId);
+        mImpl.onOverlayPackageRemoved(packageName, userId);
     }
 
     private static final class DummyState {
diff --git a/services/tests/servicestests/src/com/android/server/om/OverlayUtils.java b/services/tests/servicestests/src/com/android/server/om/OverlayUtils.java
index 9e713c6dc77..e9a2e99910a 100644
--- a/services/tests/servicestests/src/com/android/server/om/OverlayUtils.java
+++ b/services/tests/servicestests/src/com/android/server/om/OverlayUtils.java
@@ -15,10 +15,13 @@ import android.os.ServiceManager;
 import android.os.SystemClock;
 
 import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executors;
 import java.util.concurrent.LinkedBlockingQueue;
 
 class OverlayUtils {
     private static final int DELAY_MS = 100;
+    private static final int MAX_WAIT_TIME = 10 * 1000;
 
     public static void enable(@NonNull final Context ctx, @NonNull final String packageName,
             final int userId) throws Exception {
@@ -65,6 +68,38 @@ class OverlayUtils {
         return info.isEnabled();
     }
 
+    public static void pollUntilOverlayAppears(@NonNull final Context ctx,
+            @NonNull final String packageName, final int userId) throws Exception {
+        Executors
+            .newSingleThreadExecutor()
+            .submit(new Poll(ctx, packageName, userId))
+            .get(MAX_WAIT_TIME, MILLISECONDS);
+    }
+
+    private static class Poll implements Callable<Void> {
+        private final Context mContext;
+        private final String mPackageName;
+        private final int mUserId;
+
+        public Poll(@NonNull final Context ctx, @NonNull final String packageName,
+                final int userId) {
+            mContext = ctx;
+            mPackageName = packageName;
+            mUserId = userId;
+        }
+
+        @Override
+        public Void call() throws Exception {
+            final IOverlayManager om = getOverlayManager(mContext);
+            OverlayInfo oi = om.getOverlayInfo(mPackageName, mUserId);
+            while (oi == null) {
+                SystemClock.sleep(DELAY_MS);
+                oi = om.getOverlayInfo(mPackageName, mUserId);
+            }
+            return null;
+        }
+    }
+
     private static IOverlayManager getOverlayManager(@NonNull final Context ctx) {
         final IBinder b = ServiceManager.getService(Context.OVERLAY_SERVICE);
         return IOverlayManager.Stub.asInterface(b);
@@ -82,8 +117,6 @@ class OverlayUtils {
     private OverlayUtils() {}
 
     private static class IntentListener extends BroadcastReceiver implements AutoCloseable {
-        private static final int MAX_WAIT_TIME = 30 * 1000;
-
         private final BlockingQueue<Integer> mResults = new LinkedBlockingQueue<Integer>(1);
         private final Context mContext;
 
diff --git a/services/tests/servicestests/src/com/android/server/om/PackageUtils.java b/services/tests/servicestests/src/com/android/server/om/PackageUtils.java
index a06822a52fa..85bdfb854f6 100644
--- a/services/tests/servicestests/src/com/android/server/om/PackageUtils.java
+++ b/services/tests/servicestests/src/com/android/server/om/PackageUtils.java
@@ -13,6 +13,7 @@ import android.content.Intent;
 import android.content.IntentFilter;
 import android.content.IntentSender;
 import android.content.pm.PackageInfo;
+import android.content.pm.PackageInstaller.SessionParams;
 import android.content.pm.PackageInstaller;
 import android.content.pm.PackageManager;
 import android.net.Uri;
@@ -30,8 +31,9 @@ class PackageUtils {
     public static void install(@NonNull final Context ctx, @NonNull final Uri uri)
             throws Exception {
         final PackageInstaller installer = ctx.getPackageManager().getPackageInstaller();
-        final int sessionId =
-                installer.createSession(new PackageInstaller.SessionParams(MODE_FULL_INSTALL));
+        SessionParams params = new SessionParams(MODE_FULL_INSTALL);
+        params.installFlags |= PackageManager.INSTALL_REPLACE_EXISTING;
+        final int sessionId = installer.createSession(params);
         final PackageInstaller.Session session = installer.openSession(sessionId);
 
         try (
diff --git a/services/tests/servicestests/src/com/android/server/om/RuntimeResourceOverlayTests.java b/services/tests/servicestests/src/com/android/server/om/RuntimeResourceOverlayTests.java
index 68cc19835df..6a0d1936c1d 100644
--- a/services/tests/servicestests/src/com/android/server/om/RuntimeResourceOverlayTests.java
+++ b/services/tests/servicestests/src/com/android/server/om/RuntimeResourceOverlayTests.java
@@ -25,6 +25,7 @@ import com.android.frameworks.servicestests.R;
 import java.io.BufferedReader;
 import java.io.InputStream;
 import java.io.InputStreamReader;
+import java.util.HashMap;
 import java.util.Locale;
 
 import org.junit.AfterClass;
@@ -46,10 +47,22 @@ public class RuntimeResourceOverlayTests {
     private static final String SOME_OTHER_APP = "com.android.rrotests.some_other_app";
     private static final String SOME_OTHER_APP_OVERLAY = "com.android.rrotests.some_other_app_overlay";
 
-    private static final String SOME_OTHER_APP_URI =
-        "android.resource://com.android.frameworks.servicestests/raw/some_other_app";
 
-    private static boolean mInitOK = false;
+    private static final String URI_PREFIX =
+        "android.resource://com.android.frameworks.servicestests/raw/";
+
+    private static final String SOME_OTHER_APP_URI = URI_PREFIX + "some_other_app";
+    private static final String APP_OVERLAY_1_V2_URI = URI_PREFIX + "app_overlay_1_v2";
+
+    private static HashMap<String, Integer> OVERLAY_PACKAGES = new HashMap<>();
+
+    static {
+        OVERLAY_PACKAGES.put(APP_OVERLAY_1, R.raw.app_overlay_1);
+        OVERLAY_PACKAGES.put(APP_OVERLAY_2, R.raw.app_overlay_2);
+        OVERLAY_PACKAGES.put(SYSTEM_OVERLAY_1, R.raw.system_overlay_1);
+        OVERLAY_PACKAGES.put(SYSTEM_OVERLAY_2, R.raw.system_overlay_2);
+        OVERLAY_PACKAGES.put(SOME_OTHER_APP_OVERLAY, R.raw.some_other_app_overlay);
+    }
 
     private Context mContext;
     private int mUserId;
@@ -60,36 +73,20 @@ public class RuntimeResourceOverlayTests {
         Context ctx = InstrumentationRegistry.getContext();
         int userId = UserHandle.myUserId();
 
-        // When support for overlays in /data is added, the checks below should
-        // be replaced with calls to PackageUtils.install.
-        try {
-            assumeTrue(PackageUtils.isInstalled(ctx, APP_OVERLAY_1));
-            assumeTrue(PackageUtils.isInstalled(ctx, APP_OVERLAY_2));
-            assumeTrue(PackageUtils.isInstalled(ctx, SYSTEM_OVERLAY_1));
-            assumeTrue(PackageUtils.isInstalled(ctx, SYSTEM_OVERLAY_2));
-            assumeTrue(PackageUtils.isInstalled(ctx, SOME_OTHER_APP_OVERLAY));
-            mInitOK = true;
-        } catch (AssumptionViolatedException e) {
-            throw new AssumptionViolatedException("Missing overlay packages: run " +
-                    "prepare-overlay-tests.sh, reboot the device and try again");
+        for (String packageName : OVERLAY_PACKAGES.keySet()) {
+            PackageUtils.install(ctx, Uri.parse(URI_PREFIX + OVERLAY_PACKAGES.get(packageName)));
+            OverlayUtils.pollUntilOverlayAppears(ctx, packageName, userId);
         }
     }
 
     @AfterClass
     public static void afterClass() throws Exception {
-        if (!mInitOK) {
-            return;
-        }
-
         Context ctx = InstrumentationRegistry.getContext();
         int userId = UserHandle.myUserId();
 
-        OverlayUtils.disable(ctx, SOME_OTHER_APP_OVERLAY, userId);
-
-        OverlayUtils.disable(ctx, SYSTEM_OVERLAY_2, userId);
-        OverlayUtils.disable(ctx, SYSTEM_OVERLAY_1, userId);
-        OverlayUtils.disable(ctx, APP_OVERLAY_2, userId);
-        OverlayUtils.disable(ctx, APP_OVERLAY_1, userId);
+        for (String packageName : OVERLAY_PACKAGES.keySet()) {
+            PackageUtils.uninstall(ctx, packageName);
+        }
     }
 
     @Before
@@ -474,6 +471,28 @@ public class RuntimeResourceOverlayTests {
         assertResource(1, R.integer.i);
     }
 
+    @Test
+    public void testUpgradeOverlay() throws Exception {
+        try {
+            OverlayUtils.enable(mContext, APP_OVERLAY_1, mUserId);
+            assertResource(1, R.integer.i);
+
+            PackageUtils.install(mContext, Uri.parse(APP_OVERLAY_1_V2_URI));
+            OverlayUtils.pollUntilOverlayAppears(mContext, APP_OVERLAY_1, mUserId);
+
+            // give the overlay manager time to detect the new package and
+            // inform the package manager about which overlays to use
+            SystemClock.sleep(1000);
+
+            assertResource(2, R.integer.i);
+        } finally {
+            PackageUtils.uninstall(mContext, APP_OVERLAY_1);
+            SystemClock.sleep(1000);
+            PackageUtils.install(mContext, Uri.parse(URI_PREFIX + R.raw.app_overlay_1));
+            OverlayUtils.pollUntilOverlayAppears(mContext, APP_OVERLAY_1, mUserId);
+        }
+    }
+
     private void assertResource(boolean expected, int resid) throws Exception {
         boolean actual = mResources.getBoolean(resid);
         assertEquals(expected, actual);
diff --git a/services/tests/servicestests/src/com/android/server/pm/PackageParserTest.java b/services/tests/servicestests/src/com/android/server/pm/PackageParserTest.java
index fe0d4f4e134..8ff4852f642 100644
--- a/services/tests/servicestests/src/com/android/server/pm/PackageParserTest.java
+++ b/services/tests/servicestests/src/com/android/server/pm/PackageParserTest.java
@@ -196,7 +196,6 @@ public class PackageParserTest {
         assertEquals(a.installLocation, b.installLocation);
         assertEquals(a.coreApp, b.coreApp);
         assertEquals(a.mRequiredForAllUsers, b.mRequiredForAllUsers);
-        assertEquals(a.mTrustedOverlay, b.mTrustedOverlay);
         assertEquals(a.use32bitAbi, b.use32bitAbi);
         assertEquals(a.packageName, b.packageName);
         assertTrue(Arrays.equals(a.splitNames, b.splitNames));
@@ -429,7 +428,6 @@ public class PackageParserTest {
         pkg.installLocation = 100;
         pkg.coreApp = true;
         pkg.mRequiredForAllUsers = true;
-        pkg.mTrustedOverlay = true;
         pkg.use32bitAbi = true;
         pkg.packageName = "foo";
         pkg.splitNames = new String[] { "foo2" };
diff --git a/services/tests/servicestests/test-apks/app_overlay_1/AndroidManifest.xml b/services/tests/servicestests/test-apks/app_overlay_1/AndroidManifest.xml
index 5cc1491602c..6a6b630e75e 100644
--- a/services/tests/servicestests/test-apks/app_overlay_1/AndroidManifest.xml
+++ b/services/tests/servicestests/test-apks/app_overlay_1/AndroidManifest.xml
@@ -1,4 +1,6 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
-    package="com.android.rrotests.app_overlay_1">
+    package="com.android.rrotests.app_overlay_1"
+    android:versionCode="1"
+    android:versionName="1.0">
     <overlay android:targetPackage="com.android.frameworks.servicestests"/>
 </manifest>
diff --git a/services/tests/servicestests/test-apks/app_overlay_1_v2/AndroidManifest.xml b/services/tests/servicestests/test-apks/app_overlay_1_v2/AndroidManifest.xml
new file mode 100644
index 00000000000..7d4fc42927b
--- /dev/null
+++ b/services/tests/servicestests/test-apks/app_overlay_1_v2/AndroidManifest.xml
@@ -0,0 +1,6 @@
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    package="com.android.rrotests.app_overlay_1"
+    android:versionCode="2"
+    android:versionName="2.0">
+    <overlay android:targetPackage="com.android.frameworks.servicestests"/>
+</manifest>
diff --git a/services/tests/servicestests/test-apks/app_overlay_1_v2/build b/services/tests/servicestests/test-apks/app_overlay_1_v2/build
new file mode 100644
index 00000000000..6de2366e9dc
--- /dev/null
+++ b/services/tests/servicestests/test-apks/app_overlay_1_v2/build
@@ -0,0 +1,24 @@
+# source this script to regenerate this package as a resource in the test app
+
+compile()
+{
+    aapt package \
+        -M AndroidManifest.xml \
+        -S res \
+        -I "${OUT}"/system/framework/framework-res.apk \
+        -F /tmp/apk \
+        -f
+}
+
+sign()
+{
+    local T=$(gettop)
+    java -Djava.library.path="$(get_abs_build_var SIGNAPK_JNI_LIBRARY_PATH)" \
+        -jar "${ANDROID_HOST_OUT}"/framework/signapk.jar \
+        "${T}"/build/target/product/security/platform.x509.pem \
+        "${T}"/build/target/product/security/platform.pk8 \
+        /tmp/apk \
+        ../../res/raw/app_overlay_1_v2
+}
+
+compile && sign && rm -f /tmp/apk
diff --git a/services/tests/servicestests/test-apks/app_overlay_1_v2/res/values/values.xml b/services/tests/servicestests/test-apks/app_overlay_1_v2/res/values/values.xml
new file mode 100644
index 00000000000..20f341a3d17
--- /dev/null
+++ b/services/tests/servicestests/test-apks/app_overlay_1_v2/res/values/values.xml
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <integer name="i">2</integer>
+</resources>
-- 
2.15.1

